### PR TITLE
Enabling library dependencies while avoid increasing diff on upstream files

### PR DESCRIPTION
**Pull request by CarlosNihelton**
This PR is quite complex and can be understood in two steps:
- Enabling Vcpkg as a way to manage dependences without committing changes on `vcxproj` and manifest files. (85eda3be4ee90d1cb9f71408709132aae19ae5b4) and
- Removing hard coded YAML in the launcher source in favor of YAML-Cpp library. (the last two commits).

Although this could be broken in two pieces it would be hard to demo the first without actually bringing dependences into the play.

The constraints I considered while working on this were:
- Avoid commit configuration changes in Visual Studio projects. XML can be very unpleasant to deal with on merging conflicts.
- CI has to be able to handle any dependency added and make sure they will be present on the final build artifacts (appx bundles).
- Leverage the current workflow that customizes pieces of the build process.

So, I changed `build.go` to append a few lines of XML to `DistroLauncher.vcxproj` and `DistroLauncher-Appx.vcxproj`. Being the modifications done post the very last tag of the XML structure, it seems safe to do verbatim. I could do the Go way, with marshalling, but that would add lots of `type struct`s to the source for no other use, besides this one.

Once prepare-build modifies those project files we can be sure that msbuild will call vcpkg to fetch and build the dependencies correctly and that any `dll`s that end up being produced will be bundled for deploy.

I made a few tests locally adding more dll libraries as dependences just to prove my theory and it worked out of the box. If you want to reproduce my tests, you can:
- Add `boost-program-options` to `DistroLauncher\vcpkg.json`
```json
{
  "$schema": "https://raw.githubusercontent.com/microsoft/vcpkg/master/scripts/vcpkg.schema.json",
  "name": "ubuntu-distro-launcher",
  "version": "0.0.1",
  "dependencies": [
    "yaml-cpp",
    "expected-lite",
    "boost-program-options",
  ]
}
```
 Edit `DistroLauncher\DistroLauncher.cpp` to the following:
 ```diff
 diff --git a/DistroLauncher/DistroLauncher.cpp b/DistroLauncher/DistroLauncher.cpp
index 83828f2..e140477 100644
--- a/DistroLauncher/DistroLauncher.cpp
+++ b/DistroLauncher/DistroLauncher.cpp
@@ -79,7 +79,7 @@ int wmain(int argc, wchar_t const *argv[])
 {
     // Update the title bar of the console window.
     SetConsoleTitleW(DistributionInfo::WindowTitle.c_str());
-
+    boost::program_options::wcommand_line_parser(argc,argv);
     // Initialize a vector of arguments.
     std::vector<std::wstring_view> arguments;
     for (int index = 1; index < argc; index += 1) {
```
Rebuild everything up to the app package. You should find some version of `boost_program_options.dll` inside the appx archive, where in the screenshot below you can see `yaml-cpp.dll`. That library brings other dependencies, but they will not appear in the final result unless you really use it in code. Just creating the command line parser will not be enough.

image

Yet, notice that I changed the `build-wsl.yaml` workflow to bring the latest `vcpkg` to the build process. That is necessary to ensure the CI will fetch and build (or rescue from a cache) the dependencies in the same way as of the development environments. I'll leave an artifact built per the latest changes as if it was run by the CI so we can inspect and make sure everything is working as expected. Before creating this branch and PR, I created an experimental one just for that purpose.

I hope you really enjoy this as much as I enjoyed building it. 😄

### CODE REVIEW

**Comment by CarlosNihelton** [comment]
The build I said I'd leave. 
https://github.com/ubuntu/WSL/actions/runs/1496435774

**Comment by didrocks** [comment]
Waow, this is quite some work, well done!

Ok, the C++ part is a little bit too advanced to me I must say, so I scanned through it but I can’t ensure this is without quirks :)
I only noticed in this part some issues with alignment.

On the Go side, I think it’s ok, but shouldn’t we add that directly to the real xml files (as this is common for every release) rather than post-processing something which is constant? Of course, this adds a diff compared to upstream, but I think this is more reliable, maintainable and at least, explicit that we add this once we merge back. What do you think?

I ask for some peer-reviewing on the C++ side, but we are all more Qt-C++ than pure C++ here :p

**Comment by CarlosNihelton** [comment]
> Waow, this is quite some work, well done!
>

Thank you.


> On the Go side, I think it’s ok, but shouldn’t we add that directly to the real xml files (as this is common for every release) rather than post-processing something which is constant? Of course, this adds a diff compared to upstream, but I think this is more reliable, maintainable and at least, explicit that we add this once we merge back. What do you think?
> 

I agree that it is more reliable and explicit, have my doubts about maintainability, but the upstream project files won't change that often and those additions are really small. When I started investigating this approach, I was expecting something more verbose, with a lot off additions to the XML structure, but it turned to be much simpler, so much that I wouldn't be worried to commit those changes, actually. I just kept the Go modifications because they were already done :smile:

**Comment by CarlosNihelton** [comment]
> Ok, the C++ part is a little bit too advanced to me I must say, so I scanned through it but I can’t ensure this is without quirks :) I only noticed in this part some issues with alignment.
> 

The alignment issue was caused by a mix of tabs and spaces. I adjusted the Visual Studio settings to ensure tabs are avoided.

**Comment by didrocks** [comment]
Apart from the last misalignement standing, this is looking good to me. /!\ again, your level of C++ is higher than mine and may have missed some parts, but it looks good and I trust your testing :) Approving (you should maybe squash the commits before merging to avoid the addition + removal then?)

**Review by didrocks — APPROVED** [review]

File: DistroLauncher/WindowsUserInfo.cpp [hunk 1]
```diff
@@ -40,26 +40,47 @@ namespace DistributionInfo {
 		// because of such small feature, which would be an overkill. Shall the need
 		// for more YAML manipulation in the DistroLauncher arise, thus function should
 		// be changed to use a proper YAML manipulation library, such as yaml-cpp.
-		std::wstring WindowsUserInfo::toYaml() const {
-			std::wstring fullYaml;
-
+		std::string WindowsUserInfo::toYamlUtf8() const {
+			// yaml-cpp doesn't support wide chars.
+			std::unordered_map<std::string,
+				std::unordered_map<std::string, std::string>> sections;
```
[the remark this answers is not in the document]
Comment (CarlosNihelton) [reply DistroLauncher/WindowsUserInfo.cpp:46]: Same thing: tabs. I'll remove the tabs all around in that file. There are other files affected as well, but I'll address those in a dedicated PR. I'll take some time to find the best settings in clang-format to mimic the upstream files and I'll kill those tabs and other issues all at once while making sure that will stay consistent, standardized and opted-in by default. Visual Studio already respects the `.clang-format` file. It just needs to exist.

File: DistroLauncher/Win32Utils.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+													  static_cast<int>(utf8str.size()), nullptr, 0);
+		if (size_needed <= 0){
+			nonstd::make_unexpected(L"MultiByteToWideChar() failed. Win32 error code: "
+                                     + std::to_wstring(GetLastError()));
+		}
+
+		std::wstring result(size_needed, 0);
+		auto convResult = MultiByteToWideChar(CP_UTF8, 0, &utf8str.at(0), static_cast<int>(utf8str.size()),
+							&result.at(0), size_needed);
+        if(convResult==0){
```
[the remark this answers is not in the document]
Comment (CarlosNihelton) [reply DistroLauncher/Win32Utils.cpp]: Not by default. But there are settings to make that happen. Also, it ships clang-format and uses the  `.clang-format` file, if found. I'll make one.

**Comment by didrocks** [comment]
This looks now good for merging to me :)

**Comment by CarlosNihelton** [comment]
Forgot to squash. I don't feel good about rewriting the `main` branch, so if you don't mind, let's leave as is.

**Comment by didrocks** [comment]
yeah, no worry!

**Pull request merged**